pkg/liqonet/conncheck/receiver.go

```diff
@@ -0,0 +1,157 @@
+	connected             bool
+	latency               time.Duration
+	lastReceivedTimestamp time.Time
+	updateCallback        UpdateFunc
+}
+
+// Receiver is a receiver for conncheck messages.
+type Receiver struct {
+	peers map[string]*Peer
+	m     sync.RWMutex
+	buff  []byte
+	conn  *net.UDPConn
+}
+
+// NewReceiver creates a new conncheck receiver.
+func NewReceiver(conn *net.UDPConn) *Receiver {
+	return &Receiver{
+		peers: make(map[string]*Peer),
+		buff:  make([]byte, buffSize),
+		conn:  conn,
+	}
+}
+
+// SendPong sends a PONG message to the given address.
+func (r *Receiver) SendPong(raddr *net.UDPAddr, msg *Msg) error {
+	msg.MsgType = PONG
+	b, err := json.Marshal(msg)
+	if err != nil {
+		return fmt.Errorf("failed to marshal msg: %w", err)
+	}
+	_, err = r.conn.WriteToUDP(b, raddr)
+	if err != nil {
+		return fmt.Errorf("failed to write to %s: %w", raddr.String(), err)
+	}
+	klog.V(8).Infof("conncheck receiver: sent a PONG -> %s", msg)
+	return nil
+}
+
+// ReceivePong receives a PONG message.
+func (r *Receiver) ReceivePong(msg *Msg) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+	if peer, ok := r.peers[msg.ClusterID]; ok {
+		now := time.Now()
+		peer.lastReceivedTimestamp = now
+		peer.latency = now.Sub(msg.TimeStamp)
+		peer.connected = true
+
+		err := peer.updateCallback(true, peer.latency, now)
+		if err != nil {
+			return fmt.Errorf("failed to update peer %s: %w", msg.ClusterID, err)
+		}
+		return nil
+	}
+	return fmt.Errorf("%s is not in the peersInfo map", msg.ClusterID)
+}
+
+// InitPeer initializes a peer.
+func (r *Receiver) InitPeer(clusterID string, updateCallback UpdateFunc) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+	if _, ok := r.peers[clusterID]; ok {
+		return fmt.Errorf("conncheck receiver: %s is already in the peersInfo map", clusterID)
+	}
+	r.peers[clusterID] = &Peer{
+		connected:             false,
+		latency:               0,
+		lastReceivedTimestamp: time.Now(),
+		updateCallback:        updateCallback,
+	}
+	return nil
+}
+
+// Run starts the receiver.
+func (r *Receiver) Run() error {
+	klog.V(8).Infof("conncheck receiver: starting")
+	for {
+		n, raddr, err := r.conn.ReadFromUDP(r.buff)
+		if err != nil {
+			return fmt.Errorf("conncheck receiver: failed to read from %s: %w", raddr.String(), err)
+		}
+		msgr := &Msg{}
+		err = json.Unmarshal(r.buff[:n], msgr)
+		if err != nil {
+			return fmt.Errorf("conncheck receiver: failed to unmarshal msg: %w", err)
+		}
+		klog.V(9).Infof("conncheck receiver: received a msg -> %s", msgr)
+		switch msgr.MsgType {
+		case PING:
+			klog.V(8).Infof("conncheck receiver: received a PING %s -> %s", raddr, msgr)
+			err = r.SendPong(raddr, msgr)
+		case PONG:
+			klog.V(8).Infof("conncheck receiver: received a PONG from %s  -> %s", raddr, msgr)
+			err = r.ReceivePong(msgr)
+		}
+		if err != nil {
+			klog.Errorf("conncheck receiver: %v", err)
+		}
+	}
+}
+
+// RunDisconnectObserver starts the disconnect observer.
+func (r *Receiver) RunDisconnectObserver() error {
```

Comment:
Given error is always nil (and cannot be managed since run as go routine), it would be better to drop it completely. Same for the calling functions.